WordPress/openverse-catalog
# Use Airflow variable to omit DAGs from any Slack notification

## Fixes



## Description


This is an extension of #643, which added an Airflow variable (`silenced_slack_alerts`) to conditionally omit DAGs from Slack alerts. This PR extends that logic to allow omitting DAGs from notifications (rather than just alerts).

The driving use case for this one is the backfill of Europeana (discussed in #380). Now that we have enabled `catchup` on this DAG, when we turn it on in production it will immediately start backfilling years of missing DagRuns. This will cause our notifications Slack channel to be flooded with messages for each of these runs. We'd like to be able to be able to temporarily turn this off and on without requiring a code change each time.

## Implementation notes

Originally, I added a second Airflow variable (`silenced_slack_notifications`) that worked identically to `silenced_slack_alerts`, but for notifications. I scrapped this in favor of **changing the existing Airflow variable to allow it to skip Slack messages of any type.** The new variable is named `silenced_slack_notifications` and looks like this:

```
{
   "cleveland_museum_workflow": [
        {
            "issue": "https://github.com/WordPress/openverse-catalog/issues/229",  # An open issue
            "predicate": "KeyError"  # Silence KeyErrors  
        },
        {
            "issue": "https://github.com/WordPress/openverse-catalog/issues/438",  # A closed issue
            "predicate": "Airflow DAG Load Data Complete"  # Silence DAG completion notifications
        }
    ],
    "wikimedia_commons_workflow": [
        # ... and so on
     ]
}
```

It is a dictionary of DAG ids to lists of `SilencedSlackNotifications`, themselves a dict mapping a GitHub issue to a single `predicate` string.

* This configuration allows us to silence multiple different errors or notifications for a single DAG. While it is possible to associate them to the same GH issue, the structure encourages you to open separate issues.
* The structure is a little more verbose, but intended to be more readable and easier to type within the code for documentation purposes.
* Each `predicate` matches against Slack message body OR the message username. This allows us to do things like silence all `KeyError` alerts, or for example the notifications sent on `report_load_completion` (by targeting the user name).

## Note for Deploying
It is safe to simply merge this code, although it changes the shape of the Airflow variables, because:
* No dags are currently silenced in production
* It also changes the _name_ of the Airflow variable, so it would be safe to switch over regardless.


## Testing Instructions

Make sure you have the `slack_message_override` set to True in your Airflow variables. I used the testing configuration for Cleveland in the notes section above, to populate the Airflow `silenced_slack_notifications` variable (Admin -> Variables -> New).

You're encouraged to play around with different configurations, and adjust the testing expectations accordingly. For Cleveland, I:
* Edited the provider script to manually raise a KeyError during ingestion
* Ran the DAG
* Confirmed that while the KeyError was thrown during `pull_data`, no Slack message was sent. You can also see a message in the logs ("Skipping Slack notification for cleveland_museum_workflow")
* Confirmed that the `report_load_completion` task passed, but the slack message was not sent. This is also recorded in the logs.
* Manually edited the script to raise a different error (not KeyError) instead. Re-run the DAG and confirmed this alert is sent.
* Verify that the `report_load_completion` message IS sent for another DAG.

And now check the `check_silenced_dags` DAG:
* Run the DAG. You should see a Slack alert listing all DAGs linked to _closed_ gitHub issues. See #643 for more examples.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
